### ISSUE TITLE
irrelevant constraints and asymptotically irrelevant costs influence result

### ISSUE THREAD

**Issue by ffrohn** [draperlaboratory/kittel-koat#4]
for the following ITS, KoAT proves a quadratic upper bound (which is
sound, but imprecise):

(GOAL COMPLEXITY)
(STARTTERM (FUNCTIONSYMBOLS start))
(VAR env o70' o71' i9 i10' o1 o2 o16' o17' o5 o18' i1' o18 x)
(RULES
start(env, o70', o71') -{0,0}> f2(env, o70', o71')
f2(o1, o5, env) -{5,5}> f2(o1, o16', env) :|: o16' < o5 && o5 > 0
f2(o1, o5, env) -{4,3 + env}> f3(o1, o5, o18', env) :|: o18' <= env
f3(o1, o5, o18, env) -{7,7}> f3(o1, o5, o70', env) :|: o70' < o18 && o18
> 0 && i10' != 0
)

The loop at f2 has complexity ~o5, the transition from f3 to f2 has
complexity env, and the loop at f3 has complexity ~o18. Since o18 is
initialized with env, the complexity should be ~o5 + 2*env.

Apart from the imprecise bound, I observed two interesting things:

- we get a linear bound if we set the cost of the transition f2 -> f3 to
a constant
- we get a linear bound if we remove the (useless) constraint o10' != 0
from the last transition

**Comment by mischel** [comment]
I've looked at the proof. In the "bad" case koat does not infer that `f2 -> f3` has complexity `1`, but propagates the bound of `f2 -> f2`. This in combination with the cost of the rule gives us a quadratic bound.

The question is why no constant bound for `f2 -> f3` is generated. This should be trivially with the `constant farkas` processor.

**Comment by mischel** [comment]
When using `-smt-solver yices` I get the desired bound. This is not totally surprising different solvers use different heuristics - in this case `z3` fails to find a constant bound for `f2 -> f3`.

**Comment by ffrohn** [comment]
Ok, thanks!

**Comment by codyroux** [comment]
I'm still a bit confused by this. What heuristics are failing for Z3? If the problems are linear arithmetic, then Z3 and Yices should be doing equally well, since they're both complete on that fragment.

**Comment by mischel** [comment]
Oh. Good to know. Just know the implementation of `minismt` which is not complete for this fragment.

Indeed the problem seems to be more obnoxious. When I disable `knowledge propagation` the tool is successful (independent of the chosen solver); though for `z3` it finds the `f2 -> f3` bound at last, which should not be the case.

**Comment by mischel** [comment]
Got it. The problem seems to be `doFarkasConstant`. Running only this method gives us already a bound - it provides constant and linear bounds. Here using different solvers may change the order in which rules are oriented strictly. Together with `knowledgePropagation` and weights this explains the strange behavior.

tldr; fix implementation for farkas for constant bounds

**Comment by mischel** [comment]
This patches should do the job.

[cintexpfarkaspolo.txt](https://github.com/draperlaboratory/kittel-koat/files/161938/cintexpfarkaspolo.txt)
[cintfarkaspolo.txt](https://github.com/draperlaboratory/kittel-koat/files/161939/cintfarkaspolo.txt)

**Comment by jordanthayer** [comment]
My reading comprehension is very poor.  Applies patch suggested my mischel.  Sorry.

**Comment by codyroux** [comment]
Alright fantastic! Thanks a bunch guys!
